File: studies/fivebar/java/src/main/java/org/team100/lib/fivebar/Scenario.java
package org.team100.lib.fivebar;

/**
 * Details of a particular geometry and motor combination."""
 */
public class Scenario {
    String name;
    // link lengths, meters
    double a1;
    double a2;
    double a3;
    double a4;
    double a5;
    // position of P1, meters
    double x1;
    double y1;
    // reduction
    double ratio;
    // stall torque, Nm
    double Tmax;
    // the work envelope
    double w;
    double h;
    double xcenter;
    double ycenter;
    double xmin;
    double xmax;
    double ymin;
    double ymax;

    public double right() {
        return xcenter + w / 2;
    }

    public double left() {
        return xcenter - w / 2;
    }

    public double top() {
        return ycenter + h / 2;
    }

    public double bottom() {
        return ycenter - h / 2;
    }
}

File: studies/fivebar/python/scenario.py
from dataclasses import dataclass


@dataclass(frozen=True, kw_only=True)
class Scenario:
    """Details of a particular geometry and motor combination."""

    name: str
    # link lengths, meters
    a1: float
    a2: float
    a3: float
    a4: float
    a5: float
    # position of P1, meters
    x1: float
    y1: float
    # reduction
    ratio: float
    # stall torque, Nm
    Tmax: float
    # the work envelope
    w: float
    h: float
    xcenter: float
    ycenter: float
    xmin: float
    xmax: float
    ymin: float
    ymax: float

    def right(self) -> float:
        return self.xcenter + self.w / 2

    def left(self) -> float:
        return self.xcenter - self.w / 2

    def top(self) -> float:
        return self.ycenter + self.h / 2

    def bottom(self) -> float:
        return self.ycenter - self.h / 2
